### PR TITLE
New LA Monitoring and Reporting Template

### PR DESCRIPTION
**Pull request by adityabalaji-msft**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by adityabalaji-msft** [comment]
#sign-off

**Comment by adityabalaji-msft** [comment]
@bmoore-msft Please merge this with master if everything is fine.

File: 101-backup-la-reporting/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-backup-la-reporting%2Fazuredeploy.json" target="_blank">
+<img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.png"/>
+</a>
+<a href="http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-backup-la-reporting%2Fazuredeploy.json" target="_blank">
+<img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.png"/>
+</a>
+
+This template deploys **OMS Monitoring and Reporting solution for Azure backup** on a Log Analytics workspace. This allows you to monitor key backup parameters such as backup and restore jobs, backup alerts and Cloud storage usage across Recovery services vaults
+
+`Tags: Azure Backup, OMS Log Analytics, Monitoring`
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/README.md]: move tags to the bottom of the readme

File: 101-backup-la-reporting/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,176 @@
+{
+    "$schema":"https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion":"1.0.0.0",
+    "parameters":{
+      "workspaceLocation": {
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/azuredeploy.json]: rename this to "location" - add a defaultValue of [resourceGroup().location]

File: 101-backup-la-reporting/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+          "type":"securestring",
+          "defaultValue":"",
+          "metadata":{
+             "description":"The sasToken required to access _artifactsLocation.  When the template is deployed using the accompanying scripts, a sasToken will be automatically generated"
+          }
+       }
+    },
+    "variables":{
+       "nestedTemplates":{
+          "backup_jobs_non_log":"[concat(parameters('_artifactsLocation'), 'nestedtemplates/backup_jobs_non_log.json', parameters('_artifactsLocationSasToken'))]",
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/azuredeploy.json]: use the uri function for the uris, e.g.

"backup_jobs_non_log":"[uri(parameters('_artifactsLocation'), concat('nestedtemplates/backup_jobs_non_log.json', parameters('_artifactsLocationSasToken')))]",

File: 101-backup-la-reporting/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                      "uri":"[variables('nestedTemplates').backup_jobs_non_log]",
+                      "contentVersion":"1.0.0.0"
+                   },
+                   "parameters":{
+                     "workspaceLocation":{"value": "[parameters('workspaceLocation')]"} ,
+                     "workspaceName":{"value": "[parameters('workspaceName')]"} 
+                  
+                  }
+                   
+
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/azuredeploy.json]: format the JSON (e.g. SHIFT+ALT+F in VS Code)

File: 101-backup-la-reporting/nestedtemplates/azure_alerts.json [hunk 5]
```diff
@@ -0,0 +1,547 @@
+{
+   "$schema":"https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+   "contentVersion":"1.0.0.0",
+   "parameters":{
+      "workspaceLocation":{
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/nestedtemplates/azure_alerts.json]: update this to location in the nested templates as well

File: 101-backup-la-reporting/nestedtemplates/azure_alerts.json [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+         "type":"string",
+         "metadata":{
+            "description":"Specify the workspace name"
+         }
+      }
+   },
+   
+         "resources":[
+            {
+                "apiVersion":"2015-11-01-preview",
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/nestedtemplates/azure_alerts.json]: update apiVersion

File: 101-backup-la-reporting/nestedtemplates/backup_items.json [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+         "type":"string",
+         "metadata":{
+            "description":"Specify the workspace name"
+         }
+      }
+   },
+   
+         "resources":[
+            {
+                "apiVersion":"2015-11-01-preview",
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/nestedtemplates/backup_items.json]: update apiVersion

File: 101-backup-la-reporting/nestedtemplates/backup_jobs_log.json [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+         "metadata":{
+            "description":"Specify the workspace name"
+         }
+      }
+   },
+   
+         "resources":[
+           
+            {
+                "apiVersion":"2015-11-01-preview",
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/nestedtemplates/backup_jobs_log.json]: update apiVersion - won't comment on the rest but update those as well

File: 101-backup-la-reporting/prereqs/prereq.azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      "metadata": {
+        "description": "Specify the workspace name"
+      }
+    },
+    "workspaceLocation": {
+      "type": "string",
+      "metadata": {
+        "description": "Specify the workspace region"
+      },
+      "defaultValue": "West Europe"
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/prereqs/prereq.azuredeploy.json]: remove this and use the parameters file (you already are)

**Comment by adityabalaji-msft** [comment]
Thanks @bmoore-msft . I have made all the changes except for the api version, as of now. However, I am facing a ResourceNotFound error for the LA workspace. When I reverted back to my earlier code (which had passed the test), I got the same error.

1) To elaborate, the prereq template deployment passed (This was the template that validated the existence of pre-requisite LA workspace). Pasting the last section of the log for this step, below:

```
`DeploymentName          : prereq-deployment
ResourceGroupName       : azdo-26bd1e0e-aebc-4ee1-a6b3-360591406040-prereqs
ProvisioningState       : Succeeded
Timestamp               : 7/16/2019 10:08:29 AM
Mode                    : Incremental
TemplateLink            : 
Parameters              : 
                           Name             Type                       Value     
                           ===============  =========================  ==========
                           workspaceName    String                     ci3723bbeff1a74329940eb56a793420
                           location         String                     West Europe
                           
 Outputs                 : 
                           Name             Type                       Value     
                           ===============  =========================  ==========
                           workspaceName    String                     ci3723bbeff1a74329940eb56a793420
                           location         String                     West Europe`

```

2) However, the main template deployment was failing (for each of the nested view templates) with the below error. Pasting the error log for one view below:

```
`Resource Microsoft.Resources/deployments 'on_prem_alerts' failed with message '{
2019-07-16T10:09:47.4106919Z   "status": "Failed",
2019-07-16T10:09:47.4107420Z   "error": {
2019-07-16T10:09:47.4107843Z     "code": "ResourceDeploymentFailure",
2019-07-16T10:09:47.4108312Z     "message": "The resource operation completed with terminal provisioning state 'Failed'.",
2019-07-16T10:09:47.4108760Z     "details": [
2019-07-16T10:09:47.4109498Z       {
2019-07-16T10:09:47.4109896Z         "code": "DeploymentFailed",
2019-07-16T10:09:47.4110526Z         "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details.",
2019-07-16T10:09:47.4111041Z         "details": [
2019-07-16T10:09:47.4111622Z           {
2019-07-16T10:09:47.4112040Z             "code": "NotFound",
2019-07-16T10:09:47.4112714Z             "message": "{\r\n  \"error\": {\r\n    \"code\": \"ResourceNotFound\",\r\n    \"message\": \"The Resource 'Microsoft.OperationalInsights/workspaces/ci3723bbeff1a74329940eb56a793420' under resource group 'azdo-26bd1e0e-aebc-4ee1-a6b3-360591406040' was not found.\"\r\n  }\r\n}"
2019-07-16T10:09:47.4113436Z           }
2019-07-16T10:09:47.4113908Z         ]
2019-07-16T10:09:47.4114319Z       }
2019-07-16T10:09:47.4114904Z     ]
2019-07-16T10:09:47.4115920Z   }
2019-07-16T10:09:47.4116583Z }'`
```
3) The template deployment succeeded when I tested it manually on one of my existing workspaces (via portal)

Hence, wanted to know if this is a temporary issue, or if there is a fix I can perform from my end for the same. Thanks..

**Comment by bmoore-msft** [comment]
Do the prereqs and the resources in azuredeploy.json need to be in the same resourceGroup?

**Comment by adityabalaji-msft** [comment]
Thanks for clarifying. Yes, the prereq resources and the azuredeploy resources need to be in the same resource group

**Comment by adityabalaji-msft** [comment]
@bmoore-msft Thanks for clarifying. Yes, the prereqs (LA workspace) and the resources defined in azuredeploy.json (of type Microsoft.OperationalInsights/workspaces/views) need to be in the same resource group. 

The azuredeploy resources are essentially views/visuals that are deployed inside of the workspace (prereq)  - child resources for the workspace. So the names for these will be of the form "Microsoft.OperationalInsights/workspaces/<workspace name>/ views/<view name>". In case there is an alternate way to approach this such that the workspace and view can be part of different resource groups, please let me know.

I have made the changes requested and the template works upon manual deployment to an existing workspace. Could you please take another look and merge the commits if there are no other issues? Thanks!

File: 101-backup-la-reporting/README.md (hunk 1 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/README.md]: Updated

File: 101-backup-la-reporting/azuredeploy.json (hunk 2 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/azuredeploy.json]: Updated

File: 101-backup-la-reporting/azuredeploy.json (hunk 3 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/azuredeploy.json]: Updated

File: 101-backup-la-reporting/azuredeploy.json (hunk 4 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/azuredeploy.json]: Updated

File: 101-backup-la-reporting/nestedtemplates/azure_alerts.json (hunk 5 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/nestedtemplates/azure_alerts.json]: Updated

File: 101-backup-la-reporting/nestedtemplates/azure_alerts.json (hunk 6 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/nestedtemplates/azure_alerts.json]: Updated for this and the other below. Just a note though - the latest api version for this resource type is 2017-03-03-preview (more than a year back).  

Similarly in the prereq.azuredeployment.json, the latest version for workspace resource is "2015-11-01-preview".

File: 101-backup-la-reporting/prereqs/prereq.azuredeploy.json (hunk 9 above)
Comment (adityabalaji-msft) [reply 101-backup-la-reporting/prereqs/prereq.azuredeploy.json]: Updated

File: 101-backup-la-reporting/prereqs/prereq.azuredeploy.json [hunk 10]
```diff
@@ -3,37 +3,44 @@
...
+      "properties": {
+        "mode": "Incremental",
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+          "contentVersion": "1.0.0.0",
+          "resources": [
+            {
+              "name": "[parameters('workspaceName')]",
+              "type": "Microsoft.OperationalInsights/workspaces",
+              "apiVersion": "2015-11-01-preview",
```
Comment (bmoore-msft) [inline 101-backup-la-reporting/prereqs/prereq.azuredeploy.json]: /providers API shows the following versions available for workspaces:
            "apiVersions": [
                "2017-04-26-preview",
                "2017-03-15-preview",
                "2017-03-03-preview",
                "2017-01-01-preview",
                "2015-11-01-preview",
                "2015-03-20"
            ]
        },

Comment (adityabalaji-msft) [reply 101-backup-la-reporting/prereqs/prereq.azuredeploy.json]: Updated

**Comment by adityabalaji-msft** [comment]
@bmoore-msft Please merge this with master if it is good to go. Thanks

**Pull request merged**